osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs

```diff
@@ -225,18 +233,52 @@ private List<GroupMapping> getGroups(List<CarouselItem> items, FilterCriteria cr
                     return getGroupsBy(b => defineGroupBySource(b.BeatmapSet!.Metadata.Source), items);
 
                 case GroupMode.Collections:
-                    var collections = getCollections?.Invoke() ?? Enumerable.Empty<BeatmapCollection>();
-                    return getGroupsBy(b => defineGroupByCollection(b, collections), items);
+                {
+                    var realm = getRealm?.Invoke();
+
+                    return realm?.Run(r =>
+                    {
+                        var collections = r.All<BeatmapCollection>().AsEnumerable();
+                        return getGroupsBy(b => defineGroupByCollection(b, collections), items);
+                    }) ?? new List<GroupMapping>();
+                }
 
                 case GroupMode.MyMaps:
                     return getGroupsBy(b => defineGroupByOwnMaps(b, criteria.LocalUserId, criteria.LocalUserUsername), items);
 
+                case GroupMode.RankAchieved:
+                {
+                    var realm = getRealm?.Invoke();
+
+                    var topRankMapping = new Dictionary<BeatmapInfo, ScoreRank>(items.Count);
+
+                    return realm?.Run(r =>
+                    {
+                        var allLocalScores = r.All<ScoreInfo>()
+                                              .Filter($"{nameof(ScoreInfo.User)}.{nameof(RealmUser.OnlineID)} == $0"
+                                                      + $" && {nameof(ScoreInfo.BeatmapInfo)}.{nameof(BeatmapInfo.Hash)} == {nameof(ScoreInfo.BeatmapHash)}"
+                                                      + $" && {nameof(ScoreInfo.Ruleset)}.{nameof(RulesetInfo.ShortName)} == $1"
+                                                      + $" && {nameof(ScoreInfo.DeletePending)} == false", criteria.LocalUserId, criteria.Ruleset?.ShortName)
+                                              .OrderByDescending(s => s.TotalScore)
+                                              .ThenBy(s => s.Date);
+
+                        foreach (var score in allLocalScores)
+                        {
+                            Debug.Assert(score.BeatmapInfo != null);
+
+                            if (topRankMapping.ContainsKey(score.BeatmapInfo))
```

Comment:
Is there a reason to not use hash here?

Reply:
The issue I have here is using a *model class* as the key. I don't care if you use ID  or hash, but please use it directly.